ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -112,73 +112,24 @@ private void handleThrottlingInHeader(SdkHttpFullResponse sdkResponse) {
         // See [https://github.com/opensearch-project/ml-commons/issues/2429] for more details.
         boolean containsThrottlingException = errorsInHeader.stream().anyMatch(str -> str.startsWith("ThrottlingException"));
-        if (containsThrottlingException && executionContext.getExceptionHolder().get() == null) {
+        if (containsThrottlingException && exceptionHolder.get() == null) {
             log.error("Remote server returned error code: {}", statusCode);
-            executionContext
-                .getExceptionHolder()
+            exceptionHolder
                 .compareAndSet(
                     null,
-                    new OpenSearchStatusException(
+                    new RetryableException(
                         REMOTE_SERVICE_ERROR + "The request was denied due to remote server throttling.",
                         RestStatus.fromCode(statusCode)
                     )
                 );
         }
     }
 
-    private void processResponse(
-        Integer statusCode,
-        String body,
-        Map<String, String> parameters,
-        Map<Integer, ModelTensors> tensorOutputs
-    ) {
-        if (Strings.isBlank(body)) {
-            log.error("Remote model response body is empty!");
-            if (executionContext.getExceptionHolder().get() == null) {
-                executionContext
-                    .getExceptionHolder()
-                    .compareAndSet(null, new OpenSearchStatusException("No response from model", RestStatus.BAD_REQUEST));
-            }
-        } else {
-            if (statusCode < HttpStatus.SC_OK || statusCode > HttpStatus.SC_MULTIPLE_CHOICES) {
-                log.error("Remote server returned error code: {}", statusCode);
-                if (executionContext.getExceptionHolder().get() == null) {
-                    executionContext
-                        .getExceptionHolder()
-                        .compareAndSet(null, new OpenSearchStatusException(REMOTE_SERVICE_ERROR + body, RestStatus.fromCode(statusCode)));
-                }
-            } else {
-                try {
-                    ModelTensors tensors = processOutput(body, connector, scriptService, parameters, mlGuard);
-                    tensors.setStatusCode(statusCode);
-                    tensorOutputs.put(executionContext.getSequence(), tensors);
-                } catch (Exception e) {
-                    log.error("Failed to process response body: {}", body, e);
-                    if (executionContext.getExceptionHolder().get() == null) {
-                        executionContext
-                            .getExceptionHolder()
-                            .compareAndSet(null, new MLException("Fail to execute predict in aws connector", e));
-                    }
-                }
-            }
-        }
-    }
-
-    // Only all requests successful case will be processed here.
-    private void reOrderTensorResponses(Map<Integer, ModelTensors> tensorOutputs) {
```

Comment:
Why remove this method ? Are you trying to refactor Zan's code ?

Reply:
Yes, we need to refactor it to use GroupedActionListener. With previous code we call onResponse/onFailure only after all sub-requests have response. With the refactor, we call onResponse/onFailure for all sub-requests, and the onFailure will further trigger the retry logic for sub-request in RetryableAction.
The sort logic is moved to another ActionListener which wraps the GroupedActionListener. After all sub-requests succeed, the GroupedActionListener.onResponse will trigger the sort logic. After that the sorted results will be sent to the final delegate listener.

Reply:
LGTM, the reOrder method is basically moved to a different place which doesn't seem to break the functionality.